netket/sampler/base.py

```diff
@@ -222,24 +222,33 @@ def sample(
             sampler, machine, parameters, state=state, chain_length=chain_length
         )
 
-    @partial(jax.jit, static_argnums=(1, 4))
     def _sample_chain(
         sampler,
         machine: Union[Callable, nn.Module],
         parameters: PyTree,
         state: SamplerState,
         chain_length: int,
     ) -> Tuple[jnp.ndarray, SamplerState]:
-        _sample_next = lambda state, _: sampler.sample_next(machine, parameters, state)
+        """
+        Samples chain_length elements along the chains.
 
-        state, samples = jax.lax.scan(
-            _sample_next,
-            state,
-            xs=None,
-            length=chain_length,
-        )
+        In general this should not be overridden unless you want to modify the logic by which
+        the whole sampling is performed.
+        If using Jax, this function should be jitted
```

Comment:
So, just to make sure I understand: The inner `_sample_chain` function is explicitly `jit`ed so it is not recompiled unnecessarily for different samplers. The outer function should still be `jit`ed itself in a sampler to remove the additional function call?

Reply:
The entry point is `.sample` which calls [this](https://github.com/netket/netket/blob/cc73fa4f23595eceb88f17219670067c71873d45/netket/sampler/base.py#L375).
, so essentially .sample calls `._sample_chain`.

Since we cannot jit methods of a class (or well, we can, but we rejit them often) this must call a trampoline fucntion outside the class itself to cache jit functions.

We always have an extra function cal unfortunately.
To remove this we should make all this code functional, make the sampler itself a dataclass with no methods and use dispatch.
But another time.